apps/sales/views.py

```diff
@@ -154,4 +157,168 @@ def create(self, request, *args, **kwargs):
+
+    def retrieve(self, request, *args, **kwargs):
+        pk = kwargs.get('pk')  # Access the pk from kwargs
+        instance = self.get_object()
+        serializer = self.get_serializer(instance)
+
+        # Query SaleOrderItems model using the pk
+        items_related_data = SaleOrderItems.objects.filter(sale_order_id=pk)  # 'sale_order_id' is the FK field
+        items_related_serializer = SaleOrderItemsSerializer(items_related_data, many=True)
+
+        # get sale_order_id value from SaleOrder Instance 
+        order_id = serializer.data.get('sale_order_id')
+
+        # Query OrderAttachments model using the order_id
+        attachments_related_data = OrderAttachments.objects.filter(order_id=str(pk))
+        attachments_related_serializer = OrderAttachmentsSerializer(attachments_related_data, many=True)
+
+        # Query OrderShipments model using the order_id
+        shipments_related_data = OrderShipments.objects.filter(order_id=str(order_id))
+        shipments_related_serializer = OrderShipmentsSerializer(shipments_related_data, many=True)
+
+        # Customizing the response data
+        custom_data = {
+            "sale_order": serializer.data,
+            "sale_order_items": items_related_serializer.data,
+            "order_attachments":attachments_related_serializer.data,
+            "order_shipments": shipments_related_serializer.data,
+        }
+        return Response(custom_data)
+        
+    def put(self, request, *args, **kwargs):
+        return self.update(request, *args, **kwargs)
+        
+    def update(self, request, *args, **kwargs):
+        partial = kwargs.pop('partial', False)
+        instance = self.get_object()
+        serializer = self.get_serializer(instance, data=request.data['sale_order'], partial=partial)
+        serializer.is_valid(raise_exception=True)
+        self.perform_update(serializer)
+        data_1 = serializer.data
+
+        # Update sale_order_items 
+        sale_order_items_data = request.data.pop('sale_order_items')
+        pk = request.data['sale_order'].get('sale_order_id')
+        data_2 = update_multi_instance(sale_order_items_data,pk,SaleOrder,SaleOrderItems,SaleOrderItemsSerializer)
+
+        # Update sale_order_attachments
+        order_attachments_data = request.data.pop('order_attachments')
+        pk = request.data['sale_order'].get('sale_order_id')
+        data_3 = update_multi_instance(order_attachments_data,pk,SaleOrder,OrderAttachments,OrderAttachmentsSerializer,main_model_field_name='order_id')
+
+        # Update order_shipments
+        order_shipments_data = request.data.pop('order_shipments')
+        pk = request.data['sale_order'].get('sale_order_id')  # Fetch value from main model
+        data_4 = update_multi_instance(order_shipments_data,pk,SaleOrder,OrderShipments,OrderShipmentsSerializer,main_model_field_name='order_id')
+
+
+        if data_1 and data_2 and data_3 and data_4:  
```

Comment:
Change the data_1 and data_2 and data_3 and data_4 variables to some meaningful and understandable names.

Reply:
Updated  with meaningful names